flink-core/src/main/java/org/apache/flink/api/connector/source/lib/DataGeneratorSource.java

```diff
@@ -0,0 +1,211 @@
+ * helpful, for instance, for testing sinks that are expected to create specific metadata upon the
+ * arrival of a checkpoint barrier and other similar use cases.
+ *
+ * <p>This source is always bounded. For very long sequences (for example when the {@code count} is
+ * set to Long.MAX_VALUE), users may want to consider executing the application in a streaming
+ * manner, because, despite the fact that the produced stream is bounded, the end bound is pretty
+ * far away.
+ */
+@Experimental
+public class DataGeneratorSource<OUT>
+        implements Source<OUT, NumberSequenceSplit, Collection<NumberSequenceSplit>>,
+                ResultTypeQueryable<OUT> {
+
+    private static final long serialVersionUID = 1L;
+
+    private final SourceReaderFactory<OUT, NumberSequenceSplit> sourceReaderFactory;
+    private final TypeInformation<OUT> typeInfo;
+
+    private final NumberSequenceSource numberSource;
+
+    /**
+     * Instantiates a new {@code DataGeneratorSource}.
+     *
+     * @param generatorFunction The {@code GeneratorFunction} function.
+     * @param count The number of generated data points.
+     * @param typeInfo The type of the produced data points.
+     */
+    public DataGeneratorSource(
+            GeneratorFunction<Long, OUT> generatorFunction,
+            long count,
+            TypeInformation<OUT> typeInfo) {
+        this(generatorFunction, count, -1, typeInfo);
+    }
+
+    /**
+     * Instantiates a new {@code DataGeneratorSource}.
+     *
+     * @param generatorFunction The {@code GeneratorFunction} function.
+     * @param count The number of generated data points.
+     * @param sourceRatePerSecond The overall source rate per second (across all source subtasks).
+     * @param typeInfo The type of the produced data points.
+     */
+    public DataGeneratorSource(
+            GeneratorFunction<Long, OUT> generatorFunction,
+            long count,
+            double sourceRatePerSecond,
+            TypeInformation<OUT> typeInfo) {
+        this(
+                new GeneratorSourceReaderFactory<>(generatorFunction, sourceRatePerSecond),
+                count,
+                typeInfo);
+        ClosureCleaner.clean(
+                generatorFunction, ExecutionConfig.ClosureCleanerLevel.RECURSIVE, true);
+    }
+
+    /**
+     * Instantiates a new {@code DataGeneratorSource}. This constructor allows users can take
+     * control of the low-level data generation details by supplying a custom {@code
+     * SourceReaderFactory}. The instantiated {@code SourceReader}s are expected to produce data
+     * based on processing {@code NumberSequenceSplit}s.
+     *
+     * @param sourceReaderFactory The {@link SourceReader} factory.
+     * @param count The number of generated data points.
+     * @param typeInfo The type of the produced data points.
+     */
+    public DataGeneratorSource(
+            SourceReaderFactory<OUT, NumberSequenceSplit> sourceReaderFactory,
```

Comment:
What use-case do you have in mind for this? What would this allow you to do that you couldn't via a generator function?

Reply:
The think is that the `SourceFunction` was a very flexible interface and gave a lot of low-level control. Since we want to deprecate SourceFunction, there is a need to give similar low-level control if we expect people to migrate away from it.
Here is a concrete example from Iceberg/Apple: https://lists.apache.org/thread/9kfppzbq7r2gk5mhykc7m786sv0fc8j2
BTW, I anticipated this question ;) https://github.com/apache/flink/blob/ac87127cc8ff095f7390096853ba99aee0c7194e/flink-core/src/main/java/org/apache/flink/api/connector/source/lib/DataGeneratorSource.java#L83-L89

Reply:
That makes sense, but don't we end up with a worse version of the same problem for the future?

The `SourceReader` interface expose even more stuff than the `SourceFunction` did.
AFAICT `pollNext` (or rather the `ReaderOutput`) and `notifyCheckpointComplete` provide the full functionality of the `SourceFunction`.

If a user goes as far as implementing their own SourceReader, then to me it kinda feels like they don't need this FLIP in the first place. They could just extend the `NumberSequenceSource` and override `createReader()`.

Reply:
> That makes sense, but don't we end up with a worse version of the same problem for the future?
Which problem do you mean?

Reply:
@cachemeifyoucan Steven, since the original requirement came from you, would be great to hear your opinion on this thread.

Reply:
> @cachemeifyoucan Steven, since the original requirement came from you, would be great to hear your opinion on this thread.

That is not me. Wrong person tagged

Reply:
@cachemeifyoucan apologies for that.

Reply:
@afedulov @zentol  The ask is to be able to control the elements emitted for each checkpoint cycle. it is unclear to me how this constructor can be used to achieve that. The other option is to add checkpoint hooks/callbacks in the `GeneratorFunction` interface if it is possible.

Reply:
@stevenzwu The current constructor that accepts the GeneratorFunction wouldn't support it; I'm trying to figure out what exactly is actually required by current users.
If it's just rate-limiting by checkpoints, then we could either add a hook or maybe even build that into the generator source as well. It already has rate-limiting anyway.
In fact, if we extend the RateLimiter interface with a checkpoint complete hook this could be implemented quite easily. Maybe we should add _that_ as a constructor argument.

Reply:
@zentol the request is to be able to control the elements emitted in each checkpoint cycle so that we can assert on the expected output after each checkpoint cycle (e.g. in the sink table). https://github.com/apache/iceberg/blob/master/flink/v1.15/flink/src/test/java/org/apache/iceberg/flink/source/BoundedTestSource.java.

It is not like regular rate limiter where it typically controls message rate per second. Instead, we want a precise control of the exact messages emitted in each checkpoint cycle.

Reply:
@stevenzwu Whether a rate limiter works per-second or per-checkpoint is an implementation detail of the rate limiter (more generally speaking, the _interval_ is an implementation detail). I don't see a reason why this couldn't be implemented that way.

Reply:
> @stevenzwu Whether a rate limiter works per-second or per-checkpoint is an implementation detail of the rate limiter (more generally speaking, the _interval_ is an implementation detail). I don't see a reason why this couldn't be implemented that way.

I definitely agree with that interval can be an implementation detail. Rate limiter approach sounds good to me as long as we hook up the checkpoint cycle with `RateLimiter` or `RateLimitedSourceReader`.

Reply:
@stevenzwu , I sketched a solution, could you check if that satisfies your needs: [499a59](https://github.com/apache/flink/pull/20757/commits/499a599c56ca0cb8db4dc8988a941307ff4eb09d#diff-8300862ada04b39f51e7872e50711842aa38b875d3cd9954ae059a059467109fR45)
`DataGeneratorPerCheckpoint` produced a set of pre-defined elements per checkpoint.

Reply:
@zentol the [499a59](https://github.com/apache/flink/pull/20757/commits/499a599c56ca0cb8db4dc8988a941307ff4eb09d#diff-8300862ada04b39f51e7872e50711842aa38b875d3cd9954ae059a059467109fR45) sketch also shows why I wanted to have a constructor with the `SourceReaderFactory` parameter. You are right that users can always go for extending/wrapping the `NumberSequenceSource` themselves, but I find it much less convenient - they will end up reimplementing what we already did here. Since the main goal of this PR is to improve user experience, I though such extensibility is a good thing.

Reply:
You could've also done that by replacing the `double sourceRatePerSecond` DataGeneratorSource constructor argument with a `RateLimiter`. :)

But I'll give it some more thought, but atm, I'd still tend to make this constructor private and then re-evaluate it in the future if need be.

Reply:
I'd keep the constructor private. In the past we often did the mistake of exposing too many layers at once. Let's observe what issues users run into.
Replace the sourceRatePerSecond argument with a RateLimiter, and let's add some convenient factories for countPerSecond/-Checkpoint along with it.

Reply:
although this would again make things tricky with "global" vs "per-subtask" rates.

Reply:
> although this would again make things tricky with "global" vs "per-subtask" rates.

This could be solved by having the user pass a factory, like a "RateLimiterStrategy" (which _looks_ a bit nicer), which is then called on each subtask with the actual parallelism. The strategy would then determine the rate-per-subtask and create the rate limiter accordingly.

Reply:
Just to check that we are on the same page:
- your concern is giving too much "freedom" right away and therefore creating an implementation that is "all too powerful" and hard to evolve, similar to what happened with the `SourceFunction`
- Instead of giving users the freedom to supply the full `SourceReader` via
```
public SourceReader<OUT, NumberSequenceSource.NumberSequenceSplit> createReader(SourceReaderContext readerContext) 
```
we could limit the exposure by instead asking for 
```
public RateLimiter createRateLimiter(int parallelism) 
```

Do I get it right?

Reply:
I'm thinking something like this:
```
// I only deal with a single subtask and am not even aware that there are more!
public interface RateLimiter {
    CompletionStage<Void> acquire();
    default void notifyCheckpointComplete(long checkpointId) {}
}

// I'm a factory for RateLimiters and am aware of subtasks
public interface RateLimiterStrategy {
    RateLimiter createRateLimiter(int parallelism);

    // some discoverable built-in strategies
    static RateLimiterStrategy perSecond(int recordsPerSecond) {
        return parallelism -> new GuavaRateLimiter(recordsPerSecond % parallelism);
    }

    static RateLimiterStrategy perCheckpoint(int recordsPerCheckpoint) {
        return parallelism -> new GatedRateLimiter(recordsPerCheckpoint % parallelism);
    }
}

public class DataGeneratorSource {
    public DataGeneratorSource(
        GeneratorFunction<Long, OUT> generatorFunction,
        long count,
        RateLimiterStrategy rateLimiterStrategy,
        TypeInformation<OUT> typeInfo) { ... }
    // pass rateLimiterStrategy to GeneratorSourceReaderFactory constructor and evaluate in newSourceReader
}
```

Reply:
I like this approach. Implemented.